cmd/gateway/zcn/gateway-zcn.go

```diff
@@ -520,6 +522,103 @@ func (zob *zcnObjects) PutObject(ctx context.Context, bucket, object string, r *
 	return
 }
 
+func (zob *zcnObjects) PutMultipleObjects(
+	ctx context.Context,
+	bucket string,
+	objects []string,
+	r []*minio.PutObjReader,
+	opts []minio.ObjectOptions,
+) ([]minio.ObjectInfo, []error) {
+	total := len(objects)
+	if total <= 0 {
+		return nil, []error{fmt.Errorf("no files to upload")}
+	}
+
+	if total != len(r) || total != len(opts) {
+		return nil, []error{fmt.Errorf("length mismatch of objects with file readers or with options")}
+	}
+
+	remotePaths := make([]string, total)
+	for i, object := range objects {
+		if bucket == rootBucketName {
+			remotePaths[i] = filepath.Join(rootPath, object)
+		} else {
+			remotePaths[i] = filepath.Join(rootPath, bucket, object)
+		}
+	}
+	operationRequests := make([]sdk.OperationRequest, total)
+	objectInfo := make([]minio.ObjectInfo, total)
+	cbs := make([]*statusCB, total)
+	for i := 0; i < total; i++ {
+		var ref *sdk.ORef
+		ref, err := getSingleRegularRef(zob.alloc, remotePaths[i])
+		if err != nil {
+			if !isPathNoExistError(err) {
+				return nil, []error{err}
+			}
+		}
+
+		var isUpdate bool
+		if ref != nil {
+			isUpdate = true
+		}
+
+		contentType := opts[i].UserDefined["content-type"]
+		if contentType == "" {
+			contentType = "application/octet-stream"
+		}
+
+		_, fileName := filepath.Split(remotePaths[i])
+		fileMeta := sdk.FileMeta{
+			Path:       "",
+			RemotePath: remotePaths[i],
+			ActualSize: r[i].Size(),
+			MimeType:   contentType,
+			RemoteName: fileName,
+		}
+
+		cbs[i] = &statusCB{
+			doneCh: make(chan struct{}, 1),
+			errCh:  make(chan error, 1),
+		}
+		options := []sdk.ChunkedUploadOption{
+			sdk.WithStatusCallback(cbs[i]),
+			sdk.WithEncrypt(false),
+		}
+		operationRequests[i] = sdk.OperationRequest{
+			FileMeta:      fileMeta,
+			FileReader:    newMinioReader(r[i]),
+			OperationType: constants.FileOperationInsert,
+			Opts:          options,
+		}
+		if isUpdate {
+			operationRequests[i].OperationType = constants.FileOperationUpdate
+		}
+		objectInfo[i] = minio.ObjectInfo{
+			Bucket:  bucket,
+			Name:    objects[i],
+			Size:    r[i].Size(),
+			ModTime: time.Now(),
+		}
+	}
+
+	errn := zob.alloc.DoMultiOperation(operationRequests)
```

Comment:
So didn't use StartMultiUpload because I checked its code and the signature is this: `func (a *Allocation) StartMultiUpload(workdir string, localPaths []string, fileNames []string, thumbnailPaths []string, encrypts []bool, chunkNumbers []int, remotePaths []string, isUpdate bool, status StatusCallback) error`

here it is expecting localPaths which can be confusing and difficult for zs3server. We are having fileStreams in zs3server, so used the other API.

Reply:
Okay then lets change the statusCallback to a single one, no need for multiple ones error returned from this function can be same as returned from DoMultiOperation, we take consensus on the errors in it.